### PR TITLE
Use random placement library for VCH creation

### PR DESCRIPTION
**Pull request by anchal-agrawal**
[specific ci=Group6-VIC-Machine]

This change uses the placement library to relocate the VCH appliance
VM to a suitable host after it has been created, and before it is
started. This commit uses the random host placement policy since the
ranked host policy is in the works (#7459).

Future work includes using the ranked host placement policy,
incorporating a full license check and adding an integration test for
the corresponding test plan.

Towards #7280

### CODE REVIEW

File: lib/install/management/create.go [hunk 1]
```diff
@@ -100,10 +102,69 @@ func (d *Dispatcher) createPool(conf *config.VirtualContainerHostConfigSpec, set
...
+	randomPolicy := placement.NewRandomHostPolicy()
+	if randomPolicy.CheckHost(d.op, d.appliance) {
+		// The current host of the appliance is suitable; no migration needed.
+		return nil
+	}
+
+	host, err := randomPolicy.RecommendHost(d.op, d.appliance)
+	if err != nil {
+		msg := "Unable to obtain recommended host: %s"
+		d.op.Warnf(msg, err)
```
Comment (vburenin) [inline lib/install/management/create.go]: Why do you log this error here? It seems like you could just return an error and then log it where you call this function.

File: lib/install/management/create.go [hunk 2]
```diff
@@ -100,10 +102,69 @@ func (d *Dispatcher) createPool(conf *config.VirtualContainerHostConfigSpec, set
...
+func (d *Dispatcher) startAppliance(conf *config.VirtualContainerHostConfigSpec) error {
+	defer trace.End(trace.Begin("", d.op))
 	var err error
+
+	// If DRS is disabled (e.g. in a ROBO env), find a suitable host and relocate the VCH to it.
+
+	// TODO(anchal): Add license check for ROBO once #7283 is closed.
+	// TODO(anchal): Enable this check once #7440 is merged.
+	// if conf.DRSDisabled {
+	err = d.relocateAppliance()
```
Comment (vburenin) [inline lib/install/management/create.go]: Basically, I am saying that all logging could be done here.

File: lib/install/management/create_test.go [hunk 3]
```diff
@@ -90,13 +91,16 @@ func TestMain(t *testing.T) {
 		// The other way around, if we test positive case first, the VCH data and session data are modified, so we are not able to test the negative case
 		conf2, err := validator.Validate(op, input)
 		conf2.ImageStores[0].Host = "http://non-exist"
-		testCreateAppliance(op, validator.Session, conf2, installSettings, true, t)
+		_ = testCreateAppliance(op, validator.Session, conf2, installSettings, true, t)
```
Comment (vburenin) [inline lib/install/management/create_test.go]: Why did you add it?

File: lib/install/management/create_test.go [hunk 4]
```diff
@@ -245,10 +249,27 @@ func testCreateAppliance(op trace.Operation, sess *session.Session, conf *config
...
+func testRelocateAppliance(op trace.Operation, appliance *vm.VirtualMachine, sess *session.Session, conf *config.VirtualContainerHostConfigSpec, vConf *data.InstallerData, t *testing.T) {
+	d := &Dispatcher{
+		session:   sess,
+		op:        op,
+		isVC:      sess.IsVC(),
+		appliance: appliance,
+		force:     false,
+	}
+
+	err := d.relocateAppliance()
```
Comment (vburenin) [inline lib/install/management/create_test.go]: There are multiple early exist conditions in relocateAppliance() function. How do you plan to cover them in unit tests?

File: lib/install/management/create.go (hunk 2 above)
Comment (anchal-agrawal) [reply lib/install/management/create.go]: That makes sense. Agreed that having just one error log in the caller is preferred over multiple logs in the function being called. Currently, log statements are placed in `relocateAppliance` so we can capture the exact error (as there are multiple possible error scenarios) and then log the "main" warning `"Unable to relocate VCH, attempting to use its current host for powering on"` separately.

This avoid us having to log a run-on message in the caller like:

`"Unable to relocate VCH due to: Unable to obtain recommended host: <error>"`

or

`"Unable to relocate VCH due to: Failed to get canonical name for appliance:: <error>"`

While the current approach trades number of log statements for more readability, the suggested approach does the opposite. Which should we pick?

(Also this is a WIP changeset and I'll be revising it frequently.)

File: lib/install/management/create_test.go (hunk 3 above)
Comment (anchal-agrawal) [reply lib/install/management/create_test.go]: Added since the returned appliance VM can be used in the `testRelocateAppliance` func below.

File: lib/install/management/create_test.go (hunk 4 above)
Comment (anchal-agrawal) [reply lib/install/management/create_test.go]: A lot of the details in `relocateAppliance()` are pending and will be iterated upon. For example, this changeset uses random host placement since ranked placement is still being worked on. There's also some discussion to move some of the placement/policy logic deeper into the VM methods, in which case `relocateAppliance()` wouldn't have to worry about them at all.

File: lib/install/management/create.go (hunk 2 above)
Comment (vburenin) [reply lib/install/management/create.go]: I would not call existing state of logs as easily readable ;)
If you are worried about compressing too much info into one line, you could easily log it as two lines:
```
log.op.Warn(err)
log.op.Warn("Attempting to use its current host for powering on")
```

File: lib/install/management/create.go [hunk 5]
```diff
@@ -109,10 +111,84 @@ func (d *Dispatcher) createPool(conf *config.VirtualContainerHostConfigSpec, set
...
+	// hMoref := hosts[0].Reference()
+
+	host, err := randomPolicy.RecommendHost(d.op, d.appliance)
+	if err != nil {
+		msg := "Unable to obtain recommended host: %s"
+		d.op.Warnf(msg, err)
+		return errors.Errorf(msg, err)
+	}
+
+	hMoref := host.Reference()
```
Comment (cgtexmex) [inline lib/install/management/create.go]: Is this really needed?  I don't think calling `host.Reference()` two times is going to be an issue and it would be more clear

File: lib/install/management/create.go (hunk 2 above)
Comment (anchal-agrawal) [reply lib/install/management/create.go]: Perfect, fixed. Thanks!

File: lib/install/management/create_test.go (hunk 4 above)
Comment (jzt) [reply lib/install/management/create_test.go]: Yeah pretty soon most of this code will be unneeded as all the related logic will be abstracted away under `vm.VirtualMachine.PowerOn`.
